README
WWW-Contact

WWW::Contact is a system to easily get contact and addressbook data
from web-based providers. It includes support for a variety of
well-known sites, and can be easily extended to build new custom
modules.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WWW::Contact

You can also look for information at:

    WWW::Contact Google Group
        http://groups.google.com/group/perl-www-contact

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Contact

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WWW-Contact

    CPAN Ratings
        http://cpanratings.perl.org/d/WWW-Contact

    Search CPAN
        http://search.cpan.org/dist/WWW-Contact


COPYRIGHT AND LICENCE

Copyright (C) 2008 Fayland Lam

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for WWW-Contact

0.51    2018.02.28
        - Fix support for AOL mail #12 (s-nez)

0.50    2015.09.19
        - code is not working well but the build is OK now

0.47    2011.10.27
        GoogleContactsAPI: ignore google api group_name errors (Rhesa Rozendaal)
        AOL: skip data update form

0.46    2011.09.02
        Yahoo utf8 fixes
        Hotmail fixes
        AOL fixes

0.44    2010.10.05
        Return most Google contact fields instead of just name and email by Denis Howe

0.43    2010.09.13
        add support google corporate mail with Net::DNS::Resolver by sharifulin (Анатолий Шарифулин)

0.42    2010.08.21
        supplier_args (undocumented)

0.41    2010.08.05
        Hotmail bug fixes by Ahmad Anbar

0.40    2010.07.28
        Hotmail fix

0.39    2010.07.23
        webmail.aol.com to mail.aol.com

0.38    2010.06.08
        AOL fix

0.37    2010.02.23
        MSN login fix && add live.co.uk (Reported by Gez Hall)

0.36    2010.01.27
        AOL fix

0.35    2009.12.11
        Hotmail fix

0.34    2009.12.11
        Yahoo! Account Recovery Issue

0.33    2009.12.11
        fix Yahoo!

0.32    2009.09.24
        Hotmail fix

0.31    2009.09.11
        MSN login fix

0.30    2009.09.10
        fix rediffmail

0.29    2009.09.07
        yahoo fix

0.28    2009.08.10
        AOL logout issue fix
        163 mail fix

0.27    2009.07.19
        POD improvements by Anirvan Chatterjee
        Hotmail: multi-pages contact (by Quan Sun)

0.26    2009.05.22
        Funky encoding of some non-alphanumberic chars in Hotmail names fix
            by Mark Jubenville (RT 46280)

0.25    2009.05.22
        add 'use Data::Dumper;' in POD (thanks to oalders, http://annocpan.org/~FAYLAND/WWW-Contact-0.24/lib/WWW/Contact/GoogleContactsAPI.pm#note_2307)
        Funky encoding of some non-alphanumberic chars in Hotmail names fix by OALDERS (RT 46280)

0.24    2009.05.14
        make Gmail DEPRECATED, use GoogleContactsAPI
        L<WWW::Contact::BG::Abv> By Dimitar Petrov
        L<WWW::Contact::BG::Mail> By Dimitar Petrov

0.23    2009.05.04
        add Google Contacts API implemention

0.22    2009.04.02
        fix Gmail contact name bug (RT 44729)

0.21    2009.03.23
        fix Hotmail

0.20    2009.02.01
        add Hotmail/Gmail into social_network (RT 42924)
        fix Hotmail wacky results by David Moreno (RT 42925)
        fix Hotmail "You spoke, Hotmail listened" bug by FAYLAND

0.19    2009.01.03
        no code change
        try to fix CPAN install

0.18    2008.12.23
        add Moose '0.57' as deps (VERSION added)

0.17    2008.11.08
        fix "Gmail Contacts with no email cause incorrect matchup" (RT 40756)

0.16    2008.11.02
        Bug fix on Gmail module to handle fancy names(like R@j@n).
        Clean up quotes(that surround name and email) when importing data from Rediffmail.

0.15    2008.10.30
        added WWW::Contact::Plaxo by Sachin Sebastian

0.14    2008.10.28
        added WWW::Contact::Lycos by Sachin Sebastian

0.13    2008.10.28
        added WWW::Contact::Indiatimes by Sachin Sebastian
        remove google.com check and wrong password check from test files.

0.12    2008.10.27
        Edited synopsis for all derived modules.

0.11    2008.10.27
        no change

0.10    2008.10.27
        bump version up to PAUSE

0.09    2008.10.27
        add WWW::Contact::Hotmail by Fayland

0.08    2008.10.27
        added WWW::Contact::Mail by Sachin Sebastian

0.07    2008.10.27
        add WWW::Contact::AOL by Fayland

0.06    2008.10.26
        known_supplier hash
        t/999-author.t
        add WWW::Contact::CN::163 for mail.163.com

0.05    2008.10.26
        pod tweak

0.04    2008.10.26
        add Rediffmail supplier by Sachin Sebastian, C<< <sachinjsk at cpan.org> >>

0.03    2008.10.16
        add @ymail.com @rocketmail.com support

0.02    2008.10.11
        clean code and better doc.

0.01    2008.10.3
        First version, released on an unsuspecting world.



lib/WWW/Contact.pm
package WWW::Contact;

use Class::Load;
use Moose;

our $VERSION   = '0.51';
our $AUTHORITY = 'cpan:FAYLAND';

has 'errstr'   => ( is => 'rw', isa => 'Maybe[Str]' );
has 'supplier_pattern' => (
    is  => 'rw',
    isa => 'ArrayRef',
    auto_deref => 1,
    default => sub { [] }
);
has 'known_supplier' => (
    is  => 'rw',
    isa => 'HashRef',
    auto_deref => 1,
    default => sub {
        {
            'gmail.com'      => 'GoogleContactsAPI',
            'ymail.com'      => 'Yahoo',
            'rocketmail.com' => 'Yahoo',
            'rediffmail.com' => 'Rediffmail',
            'aol.com'        => 'AOL',
            'indiatimes.com' => 'Indiatimes',
            'lycos.com'      => 'Lycos',

            # cn
            '163.com'        => 'CN::163',
            'yeah.net'       => 'CN::163',
            'netease.com'    => 'CN::163',
            'popo.163.com'   => 'CN::163',

            # Mail
            'mail.com'       => 'Mail',
            'email.com'      => 'Mail',
            'iname.com'      => 'Mail',
            'cheerful.com'   => 'Mail',
            'consultant.com' => 'Mail',
            'europe.com'     => 'Mail',
            'mindless.com'   => 'Mail',
            'earthling.com'  => 'Mail',
            'myself.com'     => 'Mail',
            'techie.com'     => 'Mail',
            'usa.com'        => 'Mail',
            'writeme.com'    => 'Mail',

            # hotmail
            'hotmail.com'       => 'Hotmail',
            'live.com'          => 'Hotmail',
            'compaq.net'        => 'Hotmail',
            'hotmail.co.jp'     => 'Hotmail',
            'hotmail.co.uk'     => 'Hotmail',
            'hotmail.de'        => 'Hotmail',
            'hotmail.fr'        => 'Hotmail',
            'hotmail.it'        => 'Hotmail',
            'messengeruser.com' => 'Hotmail',
            'msn.com'           => 'Hotmail',
            'passport.com'      => 'Hotmail',
            'webtv.net'         => 'Hotmail',
            'live.co.uk'        => 'Hotmail',

            # bg
            'abv.bg'            => 'BG::Abv',
            'gbg.bg'            => 'BG::Abv',
            'gyuvectch.bg'      => 'BG::Abv',
            'mail.bg'           => 'BG::Mail',
        }
    }
);

has 'social_network' => (
    is  => 'rw',
    isa => 'HashRef',
    auto_deref => 1,
    default => sub {
        {
            # Social networks.
            'plaxo'    => 'Plaxo',
            'Hotmail'  => 'Hotmail',
            'Gmail'    => 'Gmail', # YYY? use GoogleContactsAPI?
        }
    }
);

has 'supplier_args' => (
    is  => 'rw',
    isa => 'HashRef',
    default => sub { {} }
);

has 'resolve' => (
    is  => 'rw',
    isa => 'HashRef',
    default => sub { {} }
);

has 'resolve_domain' => (
    is      => 'ro',
    isa     => 'Net::DNS::Resolver',
    lazy    => 1,
    default => sub {
        require Net::DNS::Resolver;
        Net::DNS::Resolver->new;
    },
);

sub get_contacts {
    my $self = shift;
    my ( $email, $password, $social_network ) = @_;

    unless ( $email and $password ) {
        $self->errstr('Both email and password are required.');
        return;
    }

    unless ( $email =~ m/^(.+)\@(([^.]+)\.(.+))$/ ) {
        $self->errstr('You must supply full email address.');
        return;
    }

    # get supplier module
    my $supplier;
    if($social_network) {
        $social_network = lc($social_network);
        $supplier = $self->get_supplier_by_socialnetwork($social_network);
    } else {
        $supplier = $self->get_supplier_by_email($email);
    }
    unless ($supplier) {
        if($social_network) {
            $self->errstr("$social_network is not supported yet.");
        } else {
            $self->errstr("$email is not supported yet.");
        }
        return;
    }

    my $module = 'WWW::Contact::' . $supplier;
    Class::Load::load_class($module);
    my $wc = $module->new( $self->supplier_args );

    # reset
    $self->errstr(undef);

    my $contacts = $wc->get_contacts( $email, $password );

    if ( $wc->errstr ) {
        $self->errstr( $wc->errstr );
        return;
    } else {
        return wantarray ? @$contacts : $contacts;
    }
}

sub get_supplier_by_email {
    my ($self, $email) = @_;

    my %known_supplier = $self->known_supplier;

    my ($username, $domain) = split('@', $email);

    if ( exists $known_supplier{ $domain } ) {
        return $known_supplier{ $domain };
    }

    # @yahoo.com @yahoo.XX @XX.yahoo.XX
    if ( $email =~ /[\@\.]yahoo\./ ) {
        return 'Yahoo';
    }

    my @supplier_pattern = $self->supplier_pattern;
    foreach my $supplier (@supplier_pattern) {
        my $pattern = $supplier->{pattern};
        my $mtype   = ref($pattern);
        if ( $mtype eq 'Regexp' and $email =~ $pattern ) {
            return $supplier->{supplier};
        } elsif ( $domain eq $pattern ) {
            return $supplier->{supplier};
        }
    }

    # resolve domain
    my $r = $self->resolve;
    return $r->{ $domain } if exists $r->{ $domain };

    # warn 'resolve domain';
    foreach my $mx ($self->resolve_domain->query($domain, 'MX')) {
        next unless $mx;
        for ($mx->answer) {
            # google corporate mail
            return $r->{ $domain } = $known_supplier{'gmail.com'} if $_->exchange =~ /google(?:mail)?\.com$/i;
        }
    }

    return;
}

sub get_supplier_by_socialnetwork {
    my ($self, $social_network) = @_;

    my %social_supplier = $self->social_network;

    if ( exists $social_supplier{ $social_network } ) {
        return $social_supplier{ $social_network };
    }

    return;
}

sub register_supplier {
    my ($self, $pattern, $supplier) = @_;

    unshift @{ $self->supplier_pattern }, { pattern => $pattern, supplier => $supplier };
}

no Moose;

__PACKAGE__->meta->make_immutable;

1;
__END__

=head1 NAME

WWW::Contact - Get contacts/addressbook from Web

=head1 SYNOPSIS

    use WWW::Contact;
    use Data::Dumper;

    # Get contacts from email providers.
    my $wc       = WWW::Contact->new();
    my @contacts = $wc->get_contacts('fayland@gmail.com', 'password');
    my $errstr   = $wc->errstr;
    if ($errstr) {
        die $errstr; # like 'Wrong Username or Password'
    } else {
        print Dumper(\@contacts);
    }

    # Get contacts from social networks (eg: Plaxo)
    my $ws       = WWW::Contact->new();
    # Note that the last argument for get_contacts() is mandatory,
    # or else it will try to fetch contacts from gmail.com
    my @contacts = $ws->get_contacts('itsa@gmail.com', 'password', 'plaxo');
    my $errstr   = $ws->errstr;
    if ($errstr) {
        die $errstr; # like 'Wrong Username or Password'
    } else {
        print Dumper(\@contacts);
    }


=head1 DESCRIPTION

Get contacts/addressbook from public websites

=head1 SUPPORTED EMAIL SUPPLIER

=over 4

=item Gmail

L<WWW::Contact::Gmail> by Fayland Lam, DEPRECATED for L<WWW::Contact::GoogleContactsAPI>

=item Yahoo! Mail

L<WWW::Contact::Yahoo> by Fayland Lam

=item Rediffmail

L<WWW::Contact::Rediffmail> by Sachin Sebastian

=item mail.163.com

L<WWW::Contact::CN::163> by Fayland Lam

=item AOL

L<WWW::Contact::AOL> by Fayland Lam

=item Mail

L<WWW::Contact::Mail> by Sachin Sebastian

=item Hotmail/Live Mail

L<WWW::Contact::Hotmail> by Fayland Lam

=item Indiatimes

L<WWW::Contact::Indiatimes> by Sachin Sebastian

=item Lycos

L<WWW::Contact::Lycos> by Sachin Sebastian

=item Plaxo

L<WWW::Contact::Plaxo> by Sachin Sebastian

=item GoogleContactsAPI

L<WWW::Contact::GoogleContactsAPI> by Fayland Lam, using Google Contacts Data API

=item abv.bg

L<WWW::Contact::BG::Abv> by Dimitar Petrov

=item mail.bg

L<WWW::Contact::BG::Mail> by Dimitar Petrov

=back

=head1 METHODS

=head2 register_supplier

To use custom supplier, we must register within WWW::Contact

    $wc->register_supplier( qr/\@a\.com$/, 'Unknown' );
    $wc->register_supplier( 'a.com', 'Unknown' );

The first arg is a Regexp or domain from email postfix. The second arg is the according module postfix like 'Unknown' from WWW::Contact::Unknown

=head2 get_supplier_by_email

get supplier by email.

    my $supplier = $wc->get_supplier_by_email('a@gmail.com'); # 'GoogleContactsAPI'
    my $supplier = $wc->get_supplier_by_email('a@a.com');     # 'Unknown'

=head2 get_supplier_by_socialnetwork

get supplier by social network name.

    my $supplier = $wc->get_supplier_by_socialnetwork('plaxo'); # 'Plaxo'

=head1 HOW TO WRITE YOUR OWN MODULE

Please read L<WWW::Contact::Base> and examples: L<WWW::Contact::Yahoo> and L<WWW::Contact::Plaxo>

Assuming we write a custom module as WWW::Contact::Unknown

    package WWW::Contact::Unknown;

    use Moose;
    extends 'WWW::Contact::Base';

    sub get_contacts {
        my ($self, $email, $password) = @_;

        # reset
        $self->errstr(undef);

        if ($email eq 'a@a.com' and $password ne 'a') {
            $self->errstr('Wrong Username or Password');
            return;
        }

        my @contacts = ( {
            email => 'b@b.com',
            name => 'b',
        }, {
            email => 'c@c.com',
            name => 'c'
        } );
        return wantarray ? @contacts : \@contacts;
    }

    1;

We can use it within WWW::Contact

    my $wc = WWW::Contact->new();
    $wc->register_supplier( qr/\@a\.com$/, 'Unknown' );
    # or
    # $wc->register_supplier( 'a.com', 'Unknown' );

    my @contacts = $wc->get_contacts('a@a.com', 'b');
    my $errstr = $wc->errstr;

=head1 SEE ALSO

L<WWW::Mechanize>, L<Moose>

=head1 SUPPORTS

=over 4

=item Code trunk

L<http://github.com/fayland/perl-www-contact/tree/master>

=item Group

L<http://groups.google.com/group/perl-www-contact>

=back

=head1 AUTHOR

Fayland Lam, C<< <fayland at gmail.com> >>

Sachin Sebastian, C<< <sachinjsk at cpan.org> >>

Dimitar Petrov, C<< <mitakaa at gmail.com> >>

=head1 COPYRIGHT & LICENSE

Copyright 2008 *AUTHOR* all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
